core/src/main/java/com/linecorp/armeria/internal/metric/PrometheusMetricRequestDecorator.java

```diff
@@ -0,0 +1,311 @@
+ */
+public final class PrometheusMetricRequestDecorator<T extends MetricLabel<T>,
+        I extends Request, O extends Response> {
+
+    private static final String HELP_STRING = "Armeria client/server request metric";
+
+    /**
+     * Returns a new {@link Service} decorator that tracks request stats using the Prometheus metrics  library.
+     *
+     * @param <I> Request type
+     * @param <O> Response type
+     * @param collectorRegistry Prometheus registry
+     * @param labels Labels names for metrics
+     * @param labelingFunction Must return a sorted map in natural order
+     * @return A service decorator function
+     */
+    public static <T extends MetricLabel<T>, I extends Request, O extends Response>
+    PrometheusMetricRequestDecorator<T, I, O> decorateService(
+            CollectorRegistry collectorRegistry,
+            T[] labels,
+            Function<RequestLog, Map<T, String>> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(collectorRegistry,
+                                                      labels,
+                                                      labelingFunction,
+                                                      "armeria_server");
+    }
+
+    /**
+     * Returns a new {@link Client} decorator that tracks request stats using the Prometheus metrics  library.
+     *
+     * @param <I> Request type
+     * @param <O> Response type
+     * @param prometheusRegistry Prometheus registry
+     * @param labels Labels names for metrics
+     * @param labelingFunction map of labels to values
+     * @return A client decorator function
+     */
+    public static <T extends MetricLabel<T>, I extends Request, O extends Response>
+    PrometheusMetricRequestDecorator<T, I, O> decorateClient(
+            CollectorRegistry prometheusRegistry,
+            T[] labels,
+            Function<RequestLog, Map<T, String>> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(prometheusRegistry,
+                                                      labels,
+                                                      labelingFunction,
+                                                      "armeria_client");
+    }
+
+    private final Function<RequestLog, Map<T, String>> labelingFunction;
+    private final MetricAdapter<T> metricAdapter;
+
+    /**
+     * Creates a new instance that decorates the specified {@link Service}.
+     */
+    private PrometheusMetricRequestDecorator(
+            CollectorRegistry collectorRegistry,
+            T[] metricLabels,
+            Function<RequestLog, Map<T, String>> labelingFunction,
+            String prefix) {
+        Objects.requireNonNull(collectorRegistry, "collectorRegistry");
+        Objects.requireNonNull(metricLabels, "metricLabels");
+        Objects.requireNonNull(labelingFunction, "labelingFunction");
+        Objects.requireNonNull(prefix, "prefix");
+
+        this.labelingFunction = labelingFunction;
```

Comment:
`Immutable*.copyOf` check their arguments for null, so I think you can just do `ImmutableMap.copyOf`